src/libraries/Microsoft.Extensions.DependencyInjection.Abstractions/src/ActivatorUtilities.cs

```diff
@@ -264,6 +282,77 @@ private static MethodInfo GetMethodInfo<T>(Expression<T> expr)
             return Expression.New(constructor, constructorArguments);
         }
 
+#if NETSTANDARD2_1_OR_GREATER || NETCOREAPP
+        private static ObjectFactory CreateFactoryReflection(
+            [DynamicallyAccessedMembers(DynamicallyAccessedMemberTypes.PublicConstructors)] Type instanceType,
+            Type?[] argumentTypes)
+        {
+            FindApplicableConstructor(instanceType, argumentTypes, out ConstructorInfo constructor, out int?[] parameterMap);
+
+            ParameterInfo[] constructorParameters = constructor.GetParameters();
+            FactoryParameterContext[] parameters = new FactoryParameterContext[constructorParameters.Length];
+            for (int i = 0; i < constructorParameters.Length; i++)
+            {
+                ParameterInfo constructorParameter = constructorParameters[i];
+                bool hasDefaultValue = ParameterDefaultValue.TryGetDefaultValue(constructorParameter, out object? defaultValue);
+
+                parameters[i] = new FactoryParameterContext(constructorParameter.ParameterType, hasDefaultValue, defaultValue, parameterMap[i] ?? -1);
+            }
+
+            return (IServiceProvider serviceProvider, object?[]? arguments) =>
+            {
+                object?[] constructorArguments;
+                if (parameters.Length == 0)
+                {
+                    constructorArguments = Array.Empty<object?>();
+                }
+                else
+                {
+                    constructorArguments = new object?[parameters.Length];
+                    for (int i = 0; i < parameters.Length; i++)
+                    {
+                        ref FactoryParameterContext parameter = ref parameters[i];
+                        if (parameter.ArgumentIndex != -1)
+                        {
+                            // Throws an NullReferenceException if arguments is null. Consistent with expression-based factory.
+                            constructorArguments[i] = arguments![parameter.ArgumentIndex];
+                        }
+                        else
+                        {
+                            constructorArguments[i] = GetService(
+                                serviceProvider,
+                                parameter.ParameterType,
+                                constructor.DeclaringType!,
+                                parameter.HasDefaultValue);
+                        }
+                        if (parameter.HasDefaultValue && constructorArguments[i] == null)
+                        {
+                            constructorArguments[i] = parameter.DefaultValue;
+                        }
```

Comment:
```suggestion
                        constructorArguments[i] = ((parameter.ArgumentIndex != -1) ?
                            // Throws an NullReferenceException if arguments is null. Consistent with expression-based factory.
                            arguments![parameter.ArgumentIndex]
                            :
                            GetService(
                                serviceProvider,
                                parameter.ParameterType,
                                constructor.DeclaringType!,
                                parameter.HasDefaultValue)) ?? parameter.DefaultValue;
                        }
```
Smaller code, avoid redundant `constructorArguments[i]` reads and writes.
Note that we do not need to check `parameter.HasDefaultValue` here. If the parameter does not have default value, we will replace null with null.